File: Passungstabelle/Passungstabelle_Datei.vb
﻿Imports SolidWorks.Interop.sldworks
Imports SolidWorks.Interop.swconst
Imports System.Collections.Generic

Public Class Passungstabelle_Datei
    Property Blätter As Passungstabelle_Blatt()
    Property BlätterStr As String()

    Property Attr_generell As New Dictionary(Of String, String)
    Property Attr_Übersetzungen As New Dictionary(Of String, Dictionary(Of String, String))
    Property Attr_Formate As New Dictionary(Of String, Dictionary(Of String, String))
    Property Attr_Tabelle As New Dictionary(Of String, Dictionary(Of String, String))
    '2022-06-24
    'Property Attr_Meldungen As New Dictionary(Of String, Dictionary(Of String, Boolean))

    'Property Log As New LogFile(Attr_generell)
    Property Log As LogFile

    Property Swapp As SldWorks
    Property SwDraw As DrawingDoc

    '** Speichert das aktive Blatt
    Private Property Swcsheet As SolidWorks.Interop.sldworks.Sheet

    Sub New(iSwapp As SldWorks)
        Swapp = iSwapp
    End Sub

    ' Erstellt eine Liste der Blätter in der Datei
    Function PassungsTabelleGetSheets(swmodel As SolidWorks.Interop.sldworks.ModelDoc2) As Boolean
        Dim ok As Boolean

        ok = False
        SwDraw = swmodel

        Log.WriteInfo(swmodel.GetPathName, "", False)

        '* Namen der Blätter speichern
        BlätterStr = swDraw.GetSheetNames()
        If BlätterStr.Length = 0 Then
            Log.WriteInfo(My.Resources._Keine_Blätter_in_der_Zeichnung, "", True)
            PassungsTabelleGetSheets = ok
            Exit Function
        End If

        '* Aktuelles Blatt speichern um diese später wieder zu aktivieren
        Swcsheet = SwDraw.GetCurrentSheet

        'Blätter suchen und initialisieren
        SetBlätter(swmodel)

        SwDraw.ActivateSheet(Swcsheet.GetName)
        PassungsTabelleGetSheets = True
    End Function

    Sub SetBlätter(swmodel As ModelDoc2)
        Dim i As Integer
        Dim z As Integer

        'Wenn nur die Tabelle nur auf dem ersten Blatt erscheinen soll und 
        'auf den restlichen Blättern nicht gelöscht werden soll
        'brauchen wir nur das erste Blatt untersuchen, sonst müssen auch die 
        'restlichen Blätter durchsucht werden ob eine alte Tabelle vorhanden ist
        If Attr_generell("NurAufErstemBlatt") And Attr_generell("LöschenAufRestlichenBlättern") = False Then
            z = 0
            ReDim Blätter(i)
        Else
            z = UBound(BlätterStr)
            ReDim Blätter(z)
        End If

        For i = 0 To z
            Blätter(i) = New Passungstabelle_Blatt(Swapp, Attr_generell, Attr_Übersetzungen, SwDraw.Sheet(BlätterStr(i)), swmodel, Log)
            Blätter(i).SetSheetAttr(Attr_Formate, Attr_Tabelle)
            '* Ermittlung der Ansichten auf dem Blatt
            Blätter(i).PassungsTabelleGetViews(swmodel, BlätterStr(i))
            'Blätter(i).Log = Log
        Next

        'suchen nach Passungen auf den Blättern
        PassungstabelleGetdimensionFromSheets()
    End Sub

    'Löscht die Tabellen auf allen Blättern außer dem ersten
    Sub DelTableOnOtherSheets()
        Dim sh As Sheet
        sh = SwDraw.GetCurrentSheet

        For i = 1 To BlätterStr.Length - 1
            'sh = swDraw.Sheet(BlätterStr(i))
            SwDraw.ActivateSheet(Blätter(i).Blatt.GetName)
            Blätter(i).DeleteTab()
        Next

        SwDraw.ActivateSheet(sh.GetName)
    End Sub

    'Löscht die Tabellen auf allen Blättern 
    Sub DelAllTables()
        Dim sh As Sheet
        sh = SwDraw.GetCurrentSheet

        For i = 0 To BlätterStr.Length - 1
            'sh = SwDraw.Sheet(BlätterStr(i))
            SwDraw.ActivateSheet(Blätter(i).Blatt.GetName)
            Blätter(i).DeleteTab()
        Next
        SwDraw.ActivateSheet(sh.GetName)
    End Sub

    Function PassungstabelleGetdimensionFromSheets() As Boolean
        Dim i As Integer = 0
        Dim z As Integer = 0

        'Wenn Tabelle nur auf dem ersten Blatt erscheinen soll,
        'dann wird auch nur auf dem ersten Blatt nach Passungen gesucht
        If Attr_generell("NurAufErstemBlatt") Then z = 0 Else z = Blätter.Length - 1

        For i = 0 To z
            Blätter(i).SetEinfügepunkt()
            Blätter(i).PassungsTabelleGetDimensions()
            Blätter(i).SetEinfügePunktPosition()
        Next

        PassungstabelleGetdimensionFromSheets = True
    End Function

    Sub InsertTableOnSheets()
        Dim PassungenGefunden As Boolean = False
        Dim swsheet As Sheet

        swsheet = SwDraw.GetCurrentSheet

        'Prüfung ob auf zumindest einem Blatt, Passungen gefunden wurden
        For Each blatt In Blätter
            If Not blatt.Tabelle Is Nothing Then
                If blatt.Tabelle.TabellenZeilen.Count > 0 Then
                    PassungenGefunden = True
                    'Check ob auf einem der Blätter ev. eine Tabelle vorhanden ist
                    'Wenn ja, dann wird die Tabelle gelöscht
                ElseIf Not blatt.AlteTabelle Is Nothing Then
                    SwDraw.ActivateSheet(blatt.Blatt.GetName)
                    blatt.DeleteTab()
                End If
            End If
        Next

        SwDraw.ActivateSheet(swsheet.GetName)

        'Keine Passungen gefunden
        If Not PassungenGefunden Then
            'Wenn Fehlermeldung ausgegeben werden soll
            Log.WriteInfo(My.Resources._Keine_Passungen_gefunden, "", True)
            Exit Sub
        End If


        'Wenn die Tabelle auf den restlichen Blättern gelöscht werden soll
        If Attr_generell("NurAufErstemBlatt") = True And Attr_generell("LöschenAufRestlichenBlättern") = True Then
            DelTableOnOtherSheets()
            Blätter(0).InsertTable()
        End If
        If Attr_generell("NurAufErstemBlatt") = False And Attr_generell("LöschenAufRestlichenBlättern") = False Then
            For Each n In Blätter
                n.InsertTable()
            Next
        End If
        If Attr_generell("NurAufErstemBlatt") = False And Attr_generell("LöschenAufRestlichenBlättern") = True Then
            DelAllTables()
        End If
        If Attr_generell("NurAufErstemBlatt") = True And Attr_generell("LöschenAufRestlichenBlättern") = False Then
            Blätter(0).InsertTable()
        End If

        SwDraw.ActivateSheet(Swcsheet.GetName)
    End Sub

End Class

File: Passungstabelle.CSharp/von Nahe/Passungstabelle_Datei.cs
namespace Passungstabelle.CSharp;

using System.Collections.Generic;

using SolidWorks.Interop.sldworks;


public class Passungstabelle_Datei
{
    public Passungstabelle_Blatt[] Blätter { get; set; }
    public string[] BlätterStr { get; set; }

    public GeneralSettings Attr_generell { get; set; } = new();
    public Dictionary<string, Dictionary<string, string>> Attr_Übersetzungen { get; set; } = new Dictionary<string, Dictionary<string, string>>();
    public Dictionary<string, FormatSettings> Attr_Formate { get; set; } = new();
    public Dictionary<string, TableSettings> Attr_Tabelle { get; set; } = new();

    public LogFile Log { get; set; }

    public SldWorks Swapp { get; set; }
    public DrawingDoc SwDraw { get; set; }

    private Sheet Swcsheet { get; set; }

    public Passungstabelle_Datei(SldWorks iSwapp)
    {
        this.Swapp = iSwapp;
    }

    // Erstellt eine Liste der Blätter in der Datei
    public bool PassungsTabelleGetSheets(ModelDoc2 swmodel)
    {
        bool PassungsTabelleGetSheetsRet = default;
        bool ok;

        ok = false;
        this.SwDraw = (DrawingDoc)swmodel;

        this.Log.WriteInfo(swmodel.GetPathName(), "", false);

        // * Namen der Blätter speichern
        this.BlätterStr = (string[])this.SwDraw.GetSheetNames();
        if (this.BlätterStr.Length == 0)
        {
            this.Log.WriteInfo(My.Resources._Keine_Blätter_in_der_Zeichnung, "", true);
            PassungsTabelleGetSheetsRet = ok;
            return PassungsTabelleGetSheetsRet;
        }

        // * Aktuelles Blatt speichern um diese später wieder zu aktivieren
        this.Swcsheet = (Sheet)this.SwDraw.GetCurrentSheet();

        // Blätter suchen und initialisieren
        this.SetBlätter(swmodel);

        this.SwDraw.ActivateSheet(this.Swcsheet.GetName());
        PassungsTabelleGetSheetsRet = true;
        return PassungsTabelleGetSheetsRet;
    }

    public void SetBlätter(ModelDoc2 swmodel)
    {
        var i = default(int);
        int z;

        // Wenn nur die Tabelle nur auf dem ersten Blatt erscheinen soll und 
        // auf den restlichen Blättern nicht gelöscht werden soll
        // brauchen wir nur das erste Blatt untersuchen, sonst müssen auch die 
        // restlichen Blätter durchsucht werden ob eine alte Tabelle vorhanden ist
        if (this.Attr_generell.NurAufErstemBlatt && !this.Attr_generell.LöschenAufRestlichenBlättern)
        {
            z = 0;
            this.Blätter = new Passungstabelle_Blatt[i + 1];
        }
        else
        {
            z = this.BlätterStr.Length;
            this.Blätter = new Passungstabelle_Blatt[z + 1];
        }

        var loopTo = z;
        for (i = 0; i <= loopTo; i++)
        {
            this.Blätter[i] = new Passungstabelle_Blatt(this.Swapp, this.Attr_generell, this.Attr_Übersetzungen, this.SwDraw.get_Sheet(this.BlätterStr[i]), (DrawingDoc)swmodel, this.Log);
            this.Blätter[i].SetSheetAttr(this.Attr_Formate, this.Attr_Tabelle);
            // * Ermittlung der Ansichten auf dem Blatt
            this.Blätter[i].PassungsTabelleGetViews(swmodel, this.BlätterStr[i]);
            // Blätter(i).Log = Log
        }

        // suchen nach Passungen auf den Blättern
        this.PassungstabelleGetdimensionFromSheets();
    }

    // Löscht die Tabellen auf allen Blättern außer dem ersten
    public void DelTableOnOtherSheets()
    {
        Sheet sh;
        sh = (Sheet)this.SwDraw.GetCurrentSheet();

        for (int i = 1, loopTo = this.BlätterStr.Length - 1; i <= loopTo; i++)
        {
            // sh = swDraw.Sheet(BlätterStr(i))
            this.SwDraw.ActivateSheet(this.Blätter[i].Blatt.GetName());
            this.Blätter[i].DeleteTab();
        }

        this.SwDraw.ActivateSheet(sh.GetName());
    }

    // Löscht die Tabellen auf allen Blättern 
    public void DelAllTables()
    {
        Sheet sh;
        sh = (Sheet)this.SwDraw.GetCurrentSheet();

        for (int i = 0, loopTo = this.BlätterStr.Length - 1; i <= loopTo; i++)
        {
            // sh = SwDraw.Sheet(BlätterStr(i))
            this.SwDraw.ActivateSheet(this.Blätter[i].Blatt.GetName());
            this.Blätter[i].DeleteTab();
        }
        this.SwDraw.ActivateSheet(sh.GetName());
    }

    public bool PassungstabelleGetdimensionFromSheets()
    {
        bool PassungstabelleGetdimensionFromSheetsRet = default;
        int i = 0;
        int z = 0;

        // Wenn Tabelle nur auf dem ersten Blatt erscheinen soll,
        // dann wird auch nur auf dem ersten Blatt nach Passungen gesucht
        if (this.Attr_generell.NurAufErstemBlatt)
            z = 0;
        else
            z = this.Blätter.Length - 1;

        var loopTo = z;
        for (i = 0; i <= loopTo; i++)
        {
            this.Blätter[i].SetEinfügepunkt();
            this.Blätter[i].PassungsTabelleGetDimensions();
            this.Blätter[i].SetEinfügePunktPosition();
        }

        PassungstabelleGetdimensionFromSheetsRet = true;
        return PassungstabelleGetdimensionFromSheetsRet;
    }

    public void InsertTableOnSheets()
    {
        bool PassungenGefunden = false;
        Sheet swsheet;

        swsheet = (Sheet)this.SwDraw.GetCurrentSheet();

        // Prüfung ob auf zumindest einem Blatt, Passungen gefunden wurden
        foreach (var blatt in this.Blätter)
        {
            if (blatt.Tabelle is not null)
            {
                if (blatt.Tabelle.TabellenZeilen.Count > 0)
                {
                    PassungenGefunden = true;
                }
                // Check ob auf einem der Blätter ev. eine Tabelle vorhanden ist
                // Wenn ja, dann wird die Tabelle gelöscht
                else if (blatt.AlteTabelle is not null)
                {
                    this.SwDraw.ActivateSheet(blatt.Blatt.GetName());
                    blatt.DeleteTab();
                }
            }
        }

        this.SwDraw.ActivateSheet(swsheet.GetName());

        // Keine Passungen gefunden
        if (!PassungenGefunden)
        {
            // Wenn Fehlermeldung ausgegeben werden soll
            this.Log.WriteInfo(My.Resources._Keine_Passungen_gefunden, "", true);
            return;
        }


        // Wenn die Tabelle auf den restlichen Blättern gelöscht werden soll
        if (this.Attr_generell.NurAufErstemBlatt && this.Attr_generell.LöschenAufRestlichenBlättern)
        {
            this.DelTableOnOtherSheets();
            this.Blätter[0].InsertTable();
        }
        if (!this.Attr_generell.NurAufErstemBlatt && !this.Attr_generell.LöschenAufRestlichenBlättern)
        {
            foreach (var n in this.Blätter)
            {
                n.InsertTable();
            }
        }
        if (!this.Attr_generell.NurAufErstemBlatt && this.Attr_generell.LöschenAufRestlichenBlättern)
        {
            this.DelAllTables();
        }
        if (this.Attr_generell.NurAufErstemBlatt && !this.Attr_generell.LöschenAufRestlichenBlättern)
        {
            this.Blätter[0].InsertTable();
        }

        this.SwDraw.ActivateSheet(this.Swcsheet.GetName());
    }

}